IntelliTect.TestTools.Data/DatabaseFixture.cs

```diff
@@ -0,0 +1,108 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Data.Sqlite;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace IntelliTect.TestTools.Data
+{
+    public class DatabaseFixture<TDbContext> : IDisposable where TDbContext : DbContext
+    {
+        private SqliteConnection SqliteConnection { get; }
+
+        private Lazy<DbContextOptions<TDbContext>> Options { get; }
+
+        private IServiceProvider? ServiceProvider { get; set; }
+
+        public DatabaseFixture()
+        {
+            SqliteConnection = new SqliteConnection("DataSource=:memory:");
+            SqliteConnection.Open();
+
+            Options = new Lazy<DbContextOptions<TDbContext>>(() => new DbContextOptionsBuilder<TDbContext>()
+                .UseSqlite(SqliteConnection)
+                .EnableSensitiveDataLogging()
+                .UseLoggerFactory(GetLoggerFactory())
+                .Options);
+        }
+
+        public event EventHandler<ILoggingBuilder>? BeforeLoggingSetup;
+
+        private ILoggerFactory GetLoggerFactory()
+        {
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddLogging(builder =>
+            {
+                builder.AddInMemory();
+
+                BeforeLoggingSetup?.Invoke(this, builder);;
+            });
+
+            ServiceProvider = serviceCollection.BuildServiceProvider();
+
+            return ServiceProvider
+                .GetService<ILoggerFactory>();
+        }
+
+        private TDbContext CreateNewContext()
+        {
+            var constructorInfo = typeof(TDbContext)
+                .GetConstructors()
+                .Where(x =>
+                {
+                    var parameters = x.GetParameters();
+                    return parameters.Length == 1 && parameters[0].ParameterType == typeof(DbContextOptions);
+                })
+                .SingleOrDefault();
+
+            if (constructorInfo is null)
+            {
+                throw new InvalidOperationException(
+                    $"'{typeof(TDbContext)}' must contain a constructor that has a single parameter " +
+                    $"of type '{typeof(DbContextOptions)}'");
+            }
+
+            bool alreadyCreated = Options.IsValueCreated;
+
+            var db = (TDbContext) constructorInfo.Invoke(new object[]{ Options.Value });
+
+            if (!alreadyCreated)
+            {
+                db.Database.EnsureCreated();
+            }
+
+            return db;
+        }
+
+        public async Task PerformDatabaseOperation(Func<TDbContext, Task> operation)
```

Comment:
Mentioned this idea to @COsborn2, who said I should post it in case you have any thoughts, @Keboo:

> I did have one idea, which is that you could further enforce the paradigm by creating a method where you have to pass in a callback for each segment of your test (arrange/act/assert) and the context is refreshed between each. But then this felt too clunky and prescriptive.

Reply:
I assume we are talking about applying [xml docs](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/) to the public API. I think that is a great idea, especially if the goal is to distribute as a NuGet (source link would be the other good bit to add as well). In general people want to see both when interacting with a library.

Reply:
FWIW the Directory.Build.Props should add sourcelink

Reply:
@Keboo - I read your comment and assumed you responded to my other comment (albeit on the same line) — I agree with everything you just said. 👍 

But my intent was to + you regarding a different idea, which I quoted in the parent comment — specifically, a stricter API that could prevent users from shooting themselves in the foot (i.e. only calling `PerformDatabaseOperation` once in a test method). My original idea is too restrictive, I think, but perhaps you might have a better idea.

Reply:
Ah sorry, yes @COsborn2 pointed out that I messed that one up. I am not sure there is really going to be an API solution to that problem. Though in nearly all cases the various portions of the test are easily separate, the difference between each section is, at most, reflected with a comment. Because even with a restrictive API that tries to force you to pass three separate callback delegates, aside from a CR, there is nothing stopping someone from putting the entire method in a single callback. Perhaps the better solution is a rename to make it more obvious how the method is intended to be used. Just as an idea (not sure I like it), perhaps something like `WithContext` that way if someone puts their entire unit test method inside of the callback it is clear that the entire test is acting on the same context. Thoughts?

Reply:
Added XML comments, hopefully should make things more obvious.